Add `upload_build_to_apps_cdn` action

AliSoftware opened the pull request:
## What does it do?

Implements https://github.com/Automattic/apps-infra-plans/issues/189

This adds a new `upload_build_to_apps_cdn` action in our `release-toolkit` to allow us to easily upload builds to our Apps CDN, as documented in PCYsg-15tP-p2

## Manual Testing

In addition to the unit tests, this can also be tested manually (as a kind of manual integration test) with the following steps:

 - Be sure to have a dummy file to upload for the test. I for one just picked a random PNG file and zipped it to get a `demo.png.zip`, we don't really need a real binary/build in practice for this testing.
 - Copy the WP API token from our Secret Store entry ID `12546`
 - In your terminal, after having `cd` into this repo and checked out this PR's branch, call the following command—replacing the API token and file path with the appropriate values
    ```bash
    WPCOM_API_TOKEN='{copy api token here}' \
      bundle exec fastlane run upload_build_to_apps_cdn \
      site_id:239164481 product:'WordPress.com Studio' \
      platform:'Mac - Any' build_type:Alpha visibility:external \
      version:0.1.2 build_number:1338 release_notes:'This is just for testing the CDN API' \
      file_path:$HOME/Desktop/demo.png.zip
    ```
 - Verify that the upload of the build succeeds, and that it appears on [this page](https://appscdn.wordpress.com/products/wordpress-com-studio/).
 - Click the Download button and verify that the file that gets downloaded is the exact same byte-for-byte (e.g. has the same checksum) as the one you uploaded.

The build available in pgbvBD-8l-p2 is an example of me testing the action with the above manual steps.

Alternatively, you can test the same by calling the action directly from a `Fastfile` of a test project (e.g. AutoProxxy):
 - Update the `Gemfile` to point `fastlane-plugin-wpmreleasetoolkit` to this PR's branch instead of a version
 - Run `bundle install`
 - Update your `Fastfile` to either add a dummy test lane that'd call the new action, or to replace the existing call to `appcenter_upload` in an existing lane with the call to the new `upload_build_to_apps_cdn` action, providing the proper arguments. Be sure to not provide the api token as part of the parameters, since it's a secret
 - Call `export WPCOM_API_TOKEN={copy api token from secret store here}` in your Terminal, then run `bundle exec fastlane {yourtestlane}` to test the action in… action.

## Checklist before requesting a review

- [x] Run `bundle exec rubocop` to test for code style violations and recommendations.
- [x] Add Unit Tests (aka `specs/*_spec.rb`) if applicable.
- [x] Run `bundle exec rspec` to run the whole test suite and ensure all your tests pass.
- [x] Make sure you added an entry in [the `CHANGELOG.md` file](https://github.com/wordpress-mobile/release-toolkit/blob/trunk/CHANGELOG.md#trunk) to describe your changes under the appropriate existing `###` subsection of the existing `## Trunk` section.
- [x] ~If applicable, add an entry in [the `MIGRATION.md` file](https://github.com/wordpress-mobile/release-toolkit/blob/trunk/MIGRATION.md) to describe how the changes will affect the migration from the previous major version and what the clients will need to change and consider.~

AliSoftware committed:
Initial implementation for `upload_app_to_a8c_cdn` action

AliSoftware committed:
Improve `ConfigItem`s order and descriptions

AliSoftware committed:
Fix parameters in API call

The CDN-specific parameters were never supposed to be wrapped inside an `attrs` array

AliSoftware committed:
Rename `attrs` variable to `parameters`

To avoid confusion, as those request parameters are top-level parameters for our customized API CDN endpoint, not sub-keys supposed to be nested under the `attrs` parameter of the official WP API

AliSoftware committed:
Use `a8c` not `A8C` in logs

AliSoftware committed:
Make tests about req.body more resilient

AliSoftware committed:
Simplify `build_multipart_request` implementation

AliSoftware committed:
Only return the media URL

AliSoftware committed:
Clean up code

 - Remove unnecessary comments
 - Reduce valid post status values
 - Improve description to provide valid values for parameters

AliSoftware committed:
Rename the action to `upload_build_to_apps_cdn`

AliSoftware committed:
Add CHANGELOG entry

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 500 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

AliSoftware committed:
Remove useless comments

AliSoftware commented on lib/fastlane/plugin/wpmreleasetoolkit/actions/common/upload_build_to_apps_cdn.rb:
```diff
@@ -0,0 +1,267 @@
+        # Handle the response
+        case response
+        when Net::HTTPSuccess
+          json_response = JSON.parse(response.body)
+          media = json_response['media'].first
+          media_id = media['ID']
+          media_url = media['URL']
+
+          Actions.lane_context[SharedValues::APPS_CDN_UPLOADED_FILE_URL] = media_url
+          Actions.lane_context[SharedValues::APPS_CDN_UPLOADED_FILE_ID] = media_id
+
+          UI.success('Build successfully uploaded to apps CDN')
+          UI.message("Media ID: #{media_id}")
+          UI.message("Media URL: #{media_url}")
+
+          media_url
```
@hannahtinkler So far the current implementation returns the `media['URL']` of the JSON response, but that happens to be something like `"https://appscdn.wordpress.com/wp-content/uploads/2025/03/demo.png.zip"`.

This probably won't be the most useful URL to use for the return value of the action… 💭 

So I'm wondering if it wouldn't be possible / make sense for the CDN to include the sluggified URL of the build page (e.g. `hxxps://appscdn.wordpress.com/builds/wordpress-com-studio-0-1-1-1337/`) in one of the fields of the response so we could use that as a likely-more-useful return value for this fastlane action instead?

<details><summary>For refs, this is an example of what the current API response looks like</summary>

```json
{
  "media": [                                                                                 
    {                                                                                        
      "ID": 516,                                                                             
      "URL": "https://appscdn.wordpress.com/wp-content/uploads/2025/03/demo.png.zip",        
      "guid": "http://appscdn.files.wordpress.com/2025/03/demo.png.zip",                     
      "date": "2025-03-04T18:04:22+00:00",                                                   
      "post_ID": 517,                                                                        
      "author_ID": 258585586,                                                                
      "file": "demo.png.zip",                    
      "mime_type": "application/zip",            
      "extension": "zip",                        
      "title": "demo.png",                       
      "caption": "",                             
      "description": "",                         
      "alt": "",
      "icon": "https://s1.wp.com/wp-includes/images/media/archive.png",
      "size": "136.12 KB",
      "thumbnails": {
      },
      "meta": {
        "links": {
          "self": "https://public-api.wordpress.com/rest/v1.1/sites/239164481/media/516",
          "help": "https://public-api.wordpress.com/rest/v1.1/sites/239164481/media/516/help",
          "site": "https://public-api.wordpress.com/rest/v1.1/sites/239164481",
          "parent": "https://public-api.wordpress.com/rest/v1.1/sites/239164481/posts/517"
        }
      }
    }
  ]
}
```
</details>

AliSoftware replied:
Update: I've pushed new commits to include the `post_id` and `post_url` (URL derived by using the hostname from the media URL, and appending `?p=<post_id>` to it) in the values returned / exposed by the action

AliSoftware committed:
Add new `error_on_duplicate` param

Recently added to the API

AliSoftware committed:
Return a Hash instead of just the media URL

Amongst other things, this allows to also return the post_id, and the post page URL (aka the page presenting the build we just uploaded).

AliSoftware committed:
DRY hardcoded values in specs

AliSoftware committed:
Document exposed lane_context keys

dangermattic commented on lib/fastlane/plugin/wpmreleasetoolkit/actions/common/upload_build_to_apps_cdn.rb:
```diff
@@ -0,0 +1,322 @@
+          ['APPS_CDN_UPLOADED_FILE_URL', 'The URL of the uploaded file'],
+          ['APPS_CDN_UPLOADED_FILE_ID', 'The ID of the uploaded file'],
+          ['APPS_CDN_UPLOADED_POST_ID', 'The ID of the post / page created for the uploaded build'],
+          ['APPS_CDN_UPLOADED_POST_URL', 'The URL of the post / page created for the uploaded build']
```
<table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/wordpress-mobile/release-toolkit/blob/beb2940909c03d9dae471e73e2bb44b783569e2c/lib/fastlane/plugin/wpmreleasetoolkit/actions/common/upload_build_to_apps_cdn.rb#L289"></span>Style/TrailingCommaInArrayLiteral: Put a comma after the last item of a multiline array.</td>
    </tr>
  </tbody>
</table>

twstokes commented on lib/fastlane/plugin/wpmreleasetoolkit/actions/common/upload_build_to_apps_cdn.rb:
```diff
@@ -0,0 +1,322 @@
+      VALID_PLATFORMS = ['Android', 'iOS', 'Mac - Silicon', 'Mac - Intel', 'Mac - Any', 'Windows'].freeze
+
+      def self.run(params)
+        UI.message('Uploading build to apps CDN...')
```
```suggestion
        UI.message('Uploading build to Apps CDN...')
```

twstokes commented on lib/fastlane/plugin/wpmreleasetoolkit/actions/common/upload_build_to_apps_cdn.rb:
```diff
@@ -0,0 +1,322 @@
+          Actions.lane_context[SharedValues::APPS_CDN_UPLOADED_FILE_ID] = result[:media_id]
+          Actions.lane_context[SharedValues::APPS_CDN_UPLOADED_FILE_URL] = result[:media_url]
+
+          UI.success('Build successfully uploaded to apps CDN')
```
```suggestion
          UI.success('Build successfully uploaded to Apps CDN')
```

twstokes commented on lib/fastlane/plugin/wpmreleasetoolkit/actions/common/upload_build_to_apps_cdn.rb:
```diff
@@ -0,0 +1,322 @@
+
+          result
+        else
+          UI.error("Failed to upload build to apps CDN: #{response.code} #{response.message}")
```
```suggestion
          UI.error("Failed to upload build to Apps CDN: #{response.code} #{response.message}")
```

twstokes commented on lib/fastlane/plugin/wpmreleasetoolkit/actions/common/upload_build_to_apps_cdn.rb:
```diff
@@ -0,0 +1,322 @@
+        else
+          UI.error("Failed to upload build to apps CDN: #{response.code} #{response.message}")
+          UI.error(response.body)
+          UI.user_error!('Upload to apps CDN failed')
```
```suggestion
          UI.user_error!('Upload to Apps CDN failed')
```

twstokes commented on spec/upload_build_to_apps_cdn_spec.rb:
```diff
@@ -0,0 +1,434 @@
+            file_path: file_path,
+            error_on_duplicate: true
+          )
+        end.to raise_error(FastlaneCore::Interface::FastlaneError, 'Upload to apps CDN failed')
```
```suggestion
        end.to raise_error(FastlaneCore::Interface::FastlaneError, 'Upload to Apps CDN failed')
```

twstokes commented on spec/upload_build_to_apps_cdn_spec.rb:
```diff
@@ -0,0 +1,434 @@
+            build_number: test_build_number,
+            file_path: file_path
+          )
+        end.to raise_error(FastlaneCore::Interface::FastlaneError, 'Upload to apps CDN failed')
```
```suggestion
        end.to raise_error(FastlaneCore::Interface::FastlaneError, 'Upload to Apps CDN failed')
```

twstokes commented on lib/fastlane/plugin/wpmreleasetoolkit/actions/common/upload_build_to_apps_cdn.rb:
```diff
@@ -0,0 +1,322 @@
+          build_type: params[:build_type],
+          visibility: params[:visibility].to_s.capitalize,
+          platform: params[:platform],
+          resource_type: RESOURCE_TYPE,
```
I can clearly see this constant being set as a parameter, however I'm getting:

```
Uploading build to apps CDN...
[15:41:32]: Failed to upload build to apps CDN: 422
[15:41:32]: {"errors":[{"file":"rabbit.zip","error":"validation_error","message":"The resource_type field must be one of: Build."}]}

[!] Upload to apps CDN failed
```

When running the command from the "Manual Testing" section while supplying:
- The proper API token
- An updated version, build number, and release notes
- An updated file path to the file I'm trying to upload

AliSoftware replied:
Nice catch, this was due to a small refactoring I did after my initial testing, that ended up introducing an extra newline at the end of each `form-data` parameter… and I forgot to re-test afterwards.

Fixed in a283fc1.

AliSoftware committed:
Fix typo

Co-authored-by: Tanner Stokes <tanner@tannr.com>

AliSoftware committed:
Fix extra newline in multipart/form-data formatting

AliSoftware committed:
Update unit tests to ensure no extra newline in form-data parts

twstokes reviewed (approved):
LGTM @AliSoftware - testing succeeded and produced [this page](https://appscdn.wordpress.com/builds/wordpress-com-studio-0-1-3-1339/). 🚀
